Path: More ambitious step-over and break optimizations

gwicke opened the pull request:
Extend the work from #3496 to allow the safe optimization of more
complex step transitions.

- Use the actual safePDC path for short step transitions and breaks,
  currently up to 2 cutter diameters. This value is chosen to cover
  basically all typical end step-overs, including those with heavy skew.
  Extending this much further (up to the break even point for a retract &
  rapid) will need some careful thinking for multi pass paths.
- Coordinate offset tolerances with per-operation tessellation
  tolerances, to avoid tessellation artifacts messing up paths by
  causing false retracts.  Such retracts can cause entire steps near
  vertical areas to be falsely skipped, which would cause a major
  deviation from the target model.  By considering per-job tolerances, we
  allow users to safely save computational resources by computing roughing
  operations with lower precision, or selectively increase precision for
  finish passes.
- Refine the default tessellation tolerance to GeometryTolerance / 4.
  This makes sure that the job GeometryTolerance is respected by
  operation defaults.

Thank you for creating a pull request to contribute to FreeCAD! To ease integration, please confirm the following:

- [x] Branch rebased on latest master `git pull --rebase upstream master`
- [ ] Unit tests confirmed to pass by running `./bin/FreeCAD --run-test 0`
- [ ] Commit message is [well-written](https://chris.beams.io/posts/git-commit/)
- [ ] Commit message includes `issue #<id>` or `fixes #<id>` where `<id>` is the [associated MantisBT](https://freecadweb.org/wiki/tracker#GitHub_and_MantisBT) issue id if one exists

And please remember to update the Wiki with the features added or changed once this PR is merged.  
**Note**: If you don't have wiki access, then please mention your contribution on the [0.19 Changelog Forum Thread](https://forum.freecadweb.org/viewtopic.php?f=10&t=34586).

---

gwicke commented:
Some test cases:
![image](https://user-images.githubusercontent.com/1225898/83839564-36617400-a6b1-11ea-8058-caa25d75be76.png)
![image](https://user-images.githubusercontent.com/1225898/83839589-45482680-a6b1-11ea-971e-aee0bfea936a.png)
![image](https://user-images.githubusercontent.com/1225898/83839649-70cb1100-a6b1-11ea-9584-5e8e0852d3ee.png)

Russ4262 commented on src/Mod/Path/PathScripts/PathSurfaceSupport.py:
```diff
@@ -929,22 +929,26 @@ def _calculateOffsetValue(self, isHole, isVoid=False):
             if isHole is True:
                 offset = -1 * self.obj.InternalFeaturesAdjustment.Value
-                offset += self.radius + (tolrnc / 10.0)
+                offset += self.radius + ()
```
We can remove the trailing ` + ()`.  Or, it might need to become ` + tolrnc`.

gwicke replied:
Nice catch! Fixed.

To my surprise, this was not caught by flake8. Got to love the wild world of Python ;)

Russ4262 commented:
Another suggestion is to throw in a few key `PathLog.debug()` messages throughout.  These help those who are not familiar with the code, the process, the logic, etc... and in other ways, when debugging or just trying to study the code with `PathLog.debug` enabled.

gwicke commented:
> attempt to maintain specific autonomy of the various, repeated, and identifiable transitions and stages of the paths being generated by creating unique functions to handle each

I personally see both step overs and breaks in the same category, and with similar needs. I also think that we will use them more and more interchangeably as we are adding optimizations, such as processing cut regions in groups when following the linear PDC direction would result in a lot of rapids.

Do you have specific differences in mind? If there are none right now, then I would prefer to keep the code unified for ease of maintenance. We can always diverge at a later point if we find a reason to.

gwicke commented on src/Mod/Path/PathScripts/PathSurface.py:
```diff
@@ -2095,15 +2096,37 @@ def setOclCutter(self, obj, safe=False):
+        two points in the x/y plane."""
+        p1xy, p2xy = ((p1.x, p1.y), (p2.x, p2.y))
+        pdcLine = self._planarDropCutScan(pdc, p1xy, p2xy)
+        if obj.OptimizeLinearPaths:
+            pdcLine = self._optimizeLinearSegments(pdcLine)
```
This just respects the option for transition paths as well.

gwicke commented on src/Mod/Path/PathScripts/PathSurface.py:
```diff
@@ -1383,76 +1391,69 @@ def _planarMultipassProcess(self, obj, PNTS, lMax):
+        """Generate transition commands / paths between two dropcutter steps or
+        passes. When OptimizeStepOverTransitions is enabled, uses safePDC to
+        safely optimize short (~order of cutter diameter) transitions."""
+        print(p1, p2)
```
I removed the print, good catch. Transition processing is pretty low level and high volume, so I think adding a debug print in there would be too noisy and costly.

gwicke commented on src/Mod/Path/PathScripts/PathSurface.py:
```diff
@@ -249,8 +249,9 @@ def opPropertyDefaults(self, obj, job):
             'GapThreshold': 0.005,
-            'AngularDeflection': 0.25,
-            'LinearDeflection': 0.0001,
+            'AngularDeflection': 0.25, # AngularDeflection is unused
```
It has been unused for a long time (years?). I mentioned that we should remove it in the tessellation PR, but I think we should do so consistently for both waterline and 3d surface.

Russ4262 replied:
When I suggest remove `AngularDeflection`, I am referring to it as a property of the operation, and only within the scope of the PathSurface and PathSurfaceSupport modules.  If we have no calculations dependent upon the `obj.AngularDeflection` property within the targeted modules, we can delete it within the same.
I am the one who added it a while back in order to expose it to the users, so they could adjust the mesh generation settings.  If we are not using it now, let's get rid of it, unless we see it might come in to play in the near future for additional improvements to the operation. 
((For example: There is a lot of geometry selection and manipulation code in the Waterline module, cloned from 3D Surface. I left it there because I intend to apply and utilize it for Waterline, to allow face selected waterline operations.))

gwicke replied:
Waterline OCL should have the same tessellation requirements as linear OCL. I haven't checked whether it is indeed using the same tessellation code right now, but if not then we should probably make it so, and unify the preferences as well.

In any case, I think it makes sense to do so in a separate PR.

Russ4262 replied:
> Waterline OCL should have the same tessellation requirements as linear OCL. I haven't checked whether it is indeed using the same tessellation code right now, but if not then we should probably make it so, and unify the preferences as well.
Agreed.

> In any case, I think it makes sense to tackle the AngularDeflection / tessellation cleanup in a separate PR.
Agreed.

wfm ;)

Russ4262 commented:
> attempt to maintain specific autonomy of the various, repeated, and identifiable transitions and stages of the paths being generated by creating unique functions to handle each
> 
> I personally see step overs and breaks in the same category, and with similar needs. I also think that we will use them more and more interchangeably as we are adding optimizations, such as processing cut regions in groups when strictly following the linear PDC direction would result in a lot of repetitive rapids and longer job times.
I do agree that stepovers and breaks share commonalities (I would even say they are more similar than different).  However, there are differences that could benefit from handling each in a separate function. One reason is the number of unique cut patterns, that might increase later.  Another is that for breaks, given the cut pattern, one can know and predict the geometrical path of travel that must take place through the break to the landing point on the other side.  Whereas stepovers can be distant and perhaps require additional analysis to avoid collision - and this analysis potentially changes with some of the cut patterns.

> Do you have specific differences in mind? If we have no need to diverge right now, then I would prefer to keep the code unified for ease of maintenance. We can always diverge at a later point if we find a reason to.
Overall, I don't think we need to change the code structure to isolate the aforementioned task-specific functions at this time.  I like the work and think it would benefit from access to our user base for real-world use and feedback.  
.
The aforementioned organizational suggestions stem from what I have learned working with the team of PathWB devs and wanting to make the code maintainable when we authors of the code are no longer available. ... Say 20, 30, or 50 years from now.  As, I do hope those who oversee the project are considering the long-term, even through the battles just to keep this great software current, functional, and available.

Disclaimer:  I have not compiled and tested this PR.
Claimer:  Another excelent contribution, Sir!

gwicke commented:
> One reason is the number of unique cut patterns, that might increase later. Another is that for breaks, given the cut pattern, one can know and predict the geometrical path of travel that must take place through the break to the landing point on the other side.

In the end, both were and are just linear transitions between two points though. Do you see this changing?

Re breaks, is there anything that makes them necessarily shorter or collision free? I thought breaks are just breaks in a PDC path scan, which might be of arbitrary length, and can contain arbitrary model or stock collisions in between, just as other transitions. The new transition code is set up to handle both short & long distance cases.

Russ4262 commented:
> One reason is the number of unique cut patterns, that might increase later. Another is that for breaks, given the cut pattern, one can know and predict the geometrical path of travel that must take place through the break to the landing point on the other side.
> 
> In the end, both were and are just linear transitions between two points though. Do you see this changing?

In general, I do agree about breaks and transitions both being a simple point-to-point path to avoid all collision, regardless of collision type.
I think the main focus in transition differences/complexity is with the ZigZag type cut patterns.  They are unique because the `Zag` portion (the actual transition equivalent) is actually desired to be a cut in most (if not all) cases, rather than a raise-jog-lower transition.  So, as long as we can adequately address this need to keep the cutter down and avoid collision with non-removable material, we will be good.

I support this PR and what it brings!

Russ4262 commented:
Downloaded scripts.  Initiating some testing.  I will include some `LayerMode` set to `Multi-pass`, which will also test transitions a little.

Russ4262 commented:
I think we need to adjust the behavior of `OptimizeLinearPaths`.  It is not catching many colinear paths.  When disabled, I get the multitude of SampleResolution line segments as expected.  When enabled, it is not catching segments that should be combined.  I have to step away.  I'll try and get back to it today if possible.

Russ4262 commented:
Evening Wicke,
You might need to rebase this PR to re-trigger a Travis CI build.  Travis seems frozen on this PR as I type this message. States "Some checks haven't completed yet" and "Waiting for status to be reported."  These messages have been here for part of the afternoon.

gwicke commented:
Rebased, and also added a small cleanup change to use _optimizeLinearSegments in _planarSinglePassProcess. This unifies all linear segment automation in a single place.
